File: algoexpert/easy/15.palindrome_check.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// reverse string, and then strcmp()
// O(n) time
// O(n) space
bool is_palindrome(char *word) {
  bool result = false;
  char *reversed = (char *)malloc(strlen(word) * sizeof(char));
  for (int i = strlen(word) - 1, j = 0; i >= 0; i--, j++) {
    reversed[j] = word[i];
  }
  if (strcmp(word, reversed) == 0)
    result = true;
  free(reversed);
  return result;
}

// O(n) time
// O(n) space
bool is_palindrome_rec(char *word) {
  size_t length = strlen(word);
  if (length < 2)
    return true;
  if (word[0] != word[length - 1])
    return false;
  char new_word[length - 1];
  memset(new_word, 0, length - 1);
  strncpy(new_word, &word[1], length - 2);
  // printf("%s\n", new_word);
  return is_palindrome_rec(new_word);
}

bool is_palindrome_pointers(char *word) {
  int start = 0;
  int end = strlen(word) - 1;
  while (start <= end) {
    if (word[start] != word[end])
      return false;
    start++;
    end--;
  }
  return true;
}

void print_result(char *word, bool (*callback)(char *)) {
  bool result = callback(word);
  printf("\"%s\" %s a palindrome\n", word, result ? "is" : "is NOT");
}

int main() {
  print_result("abecedario", is_palindrome);
  print_result("amanaplanacanalpanama", is_palindrome);
  printf("\n");
  print_result("abecedario", is_palindrome_rec);
  print_result("amanaplanacanalpanama", is_palindrome_rec);
  printf("\n");
  print_result("abecedario", is_palindrome_pointers);
  print_result("amanaplanacanalpanama", is_palindrome_pointers);
  return 0;
}

File: algoexpert/easy/15.palindrome_check.go
package main

import (
	"fmt"
)

// TODO: isPalindromeRev Join

func isPalindromeRev(s string) bool {
	var revs string
	for i := len(s) - 1; i >= 0; i-- {
		revs += string(s[i])
	}
	return revs == s
}

func isPalindromeRecAEHelpTail(s string, i int) bool {
	j := len(s) - 1 - i
	if i >= j {
		return true
	}
	if s[i] != s[j] {
		return false
	}
	return isPalindromeRecAEHelpTail(s, i+1)
}
func isPalindromeRecAEHelp(s string, i int) bool {
	j := len(s) - 1 - i
	if i >= j {
		return true
	}
	return s[i] == s[j] && isPalindromeRecAEHelp(s, i+1)
}
func isPalindromeRecAE(s string) bool {
	return isPalindromeRecAEHelp(s, 0)
}

func isPalindromeRec(s string) bool {
	size := len(s)
	if size == 1 {
		return true
	}
	return s[0] == s[size-1] && isPalindromeRec(s[1:size-1])
}

func isPalindromeIter(s string) bool {
	left := 0
	right := len(s) - 1
	for left <= right {
		if s[left] == s[right] {
			left++
			right--
		} else {
			return false
		}
	}
	return true
}

func main() {
	f := isPalindromeRecAE
	word := "amanaplanacanalpanama"
	if f(word) {
		fmt.Printf("`%s` is a palindrome\n", word)
	} else {
		fmt.Printf("`%s` is NOT a palindrome\n", word)
	}
	word = "abecedario"
	if f(word) {
		fmt.Printf("`%s` is a palindrome\n", word)
	} else {
		fmt.Printf("`%s` is NOT a palindrome\n", word)
	}
}
